src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,200 @@
 package core.basesyntax;
 
+import java.util.Iterator;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> last = tail;
+        Node<T> newNode = new Node<>(last, value, null);
+        tail = newNode;
+        if (last == null) {
+            head = newNode;
+        } else {
+            last.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Incorrect index " + index);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            addAtTheBeginning(value);
+            return;
+        }
+        Node<T> node = findNodeByIndex(index);
+        addBeforeNode(value, node);
     }
 
     @Override
     public void addAll(List<T> list) {
+        Iterator<T> iterator = list.iterator();
+        while (iterator.hasNext()) {
+            add(iterator.next());
+        }
```

Comment:
you can use forEach loop here. It uses an iterator for collection under the hood